virtualmachine/gcp/vm.go

```diff
@@ -206,10 +336,372 @@ func (vm *VM) DeleteDisks() error {
+			Region:            regionName,
+			GatewayAddress:    subnetwork.GatewayAddress,
+			IpCidrRange:       subnetwork.IpCidrRange,
+		})
+	}
+
+	return response, nil
+}
+
+// GetMachineTypeList gets the list of available machine types (aka flavors)
+func (vm *VM) GetMachineTypeList() ([]MachineType, error) {
+	s, err := vm.getService()
+	if err != nil {
+		return nil, err
+	}
+
+	machineTypeList, err := s.getMachineTypeList()
+	if err != nil {
+		return nil, err
+	}
+
+	response := make([]MachineType, 0)
+
+	for _, machineType := range machineTypeList {
+		response = append(response, MachineType{
+			Name:                         machineType.Name,
+			Description:                  machineType.Description,
+			GuestCpus:                    machineType.GuestCpus,
+			MemoryMb:                     machineType.MemoryMb,
+			IsSharedCpu:                  &machineType.IsSharedCpu,
+			MaximumPersistentDisks:       machineType.MaximumPersistentDisks,
+			MaximumPersistentDisksSizeGb: machineType.MaximumPersistentDisksSizeGb,
+		})
+	}
+
+	return response, nil
+}
+
+// GetDiskTypeList gets the list of available disk types
+func (vm *VM) GetDiskTypeList() ([]DiskType, error) {
+	s, err := vm.getService()
+	if err != nil {
+		return nil, err
+	}
+
+	diskTypeList, err := s.getDiskTypeList()
+	if err != nil {
+		return nil, err
+	}
+
+	response := make([]DiskType, 0)
+
+	for _, diskType := range diskTypeList {
+		response = append(response, DiskType{
+			Name:              diskType.Name,
+			Description:       diskType.Description,
+			ValidDiskSize:     diskType.ValidDiskSize,
+			DefaultDiskSizeGb: diskType.DefaultDiskSizeGb,
+		})
+	}
+
+	return response, nil
+}
+
+// GetZoneList gets the list of available zones
+func (vm *VM) GetZoneList() ([]Zone, error) {
+	s, err := vm.getService()
+	if err != nil {
+		return nil, err
+	}
+
+	zoneList, err := s.getZoneList()
+	if err != nil {
+		return nil, err
+	}
+
+	response := make([]Zone, 0)
+
+	for _, zone := range zoneList {
+		regionName := convResURLToName(zone.Region)
+		response = append(response, Zone{
+			Name:        zone.Name,
+			Description: zone.Description,
+			Region:      regionName,
+			Status:      zone.Status,
+		})
+	}
+
+	return response, nil
+}
+
+// GetRegionList gets the list of available regions
+func (vm *VM) GetRegionList() ([]Region, error) {
+	s, err := vm.getService()
+	if err != nil {
+		return nil, err
+	}
+
+	regionList, err := s.getRegionList()
+	if err != nil {
+		return nil, err
+	}
+
+	response := make([]Region, 0)
+
+	zones := make([]string, 0)
+	for _, region := range regionList {
+		zones = nil
+		for _, zoneUrl := range region.Zones {
+			zones = append(zones, convResURLToName(zoneUrl))
+		}
+		response = append(response, Region{
+			Name:        region.Name,
+			Description: region.Description,
+			Zones:       zones,
+			Status:      region.Status,
+		})
+	}
+
+	return response, nil
+}
+
+// convResURLToName returns resource name from the given resource URL
```

Comment:
should this be moved to utils.go?

Reply:
Yes, will move this to utils.go. Thanks.

Reply:
Done, kindly review.